pkg/instancetype/upgrade.go

```diff
@@ -0,0 +1,219 @@
+//nolint:lll
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+package instancetype
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	instancetypeapi "kubevirt.io/api/instancetype"
+	instancetypev1alpha1 "kubevirt.io/api/instancetype/v1alpha1"
+	instancetypev1alpha2 "kubevirt.io/api/instancetype/v1alpha2"
+	instancetypev1beta1 "kubevirt.io/api/instancetype/v1beta1"
+	generatedscheme "kubevirt.io/client-go/generated/kubevirt/clientset/versioned/scheme"
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+)
+
+type UpgraderInterface interface {
+	Upgrade(*appsv1.ControllerRevision) (*appsv1.ControllerRevision, error)
+}
+
+var _ UpgraderInterface = &Upgrader{}
+
+type Upgrader struct {
+	client     kubecli.KubevirtClient
+	vmInformer cache.SharedIndexInformer
+}
+
+func NewUpgrader(client kubecli.KubevirtClient, vmInformer cache.SharedIndexInformer) *Upgrader {
+	return &Upgrader{
+		client:     client,
+		vmInformer: vmInformer,
+	}
+}
+
+func (u *Upgrader) Upgrade(original *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
+	alreadyLatest, err := isAlreadyLatestVersion(original)
+	if err != nil {
+		return nil, err
+	}
+	if alreadyLatest {
+		return original, nil
+	}
+
+	// Find the VM owner of the CR
+	vm, err := u.discoverOwner(original)
+	if err != nil {
+		return nil, err
+	}
+
+	// Upgrade the CR object and create a new CR
+	newCR, err := u.upgradeControllerRevision(vm, original)
+	if err != nil {
+		return nil, err
+	}
+
+	// Delete original CR first to avoid naming collision if CR names equal
+	if newCR.Name == original.Name {
```

Comment:
@0xFelix added this and after some initial thought today I don't think we need this.

It's working around an issue in our functional tests where I was using the current `CreateControllerRevision` implementation to create legacy `ControllerRevisions`.

This used the current naming convention and would cause naming conflicts in the functional tests when we tried to run an upgrade on unlabelled `ControllerRevisions` containing `v1beta1` objects. However these `ControllerRevisions` wouldn't use the new naming convention introduced by this PR and thus the combination IMHO is invalid.

I've updated the functional tests to produce legacy `ControllerRevisions` when needed now using the older naming convention and removing labels as these also were never present prior to `v1beta1`.